scripts/build-test-projects/src/main.rs

```diff
@@ -84,6 +48,42 @@ fn main() {
     }
 }
 
+fn build_recursively(path: &Path) -> Vec<u64> {
+    let mut summary: Vec<u64> = vec![];
+
+    for res in fs::read_dir(path).expect("failed to walk directory") {
+        let entry = match res {
+            Ok(entry) => entry,
+            _ => continue,
+        };
+        let child_path = entry.path();
+        if !child_path.is_dir() {
+            continue;
+        } else if !dir_contains_forc_manifest(&child_path) {
+            summary.extend(build_recursively(&child_path));
+        } else {
+            let output = std::process::Command::new("forc")
+                .args(["build", "--generate-logged-types", "--path"])
+                .arg(&child_path)
+                .output()
+                .expect("failed to run `forc build` for example project");
+
+            // Print output on failure so we can read it in CI.
+            let (success, checkmark, status) = if !output.status.success() {
+                io::stdout().write_all(&output.stdout).unwrap();
+                io::stdout().write_all(&output.stderr).unwrap();
+                (0, "[x]", "failed")
+            } else {
+                (1, "[✓]", "succeeded")
+            };
+            println!("  {}: {} {}!", checkmark, child_path.display(), status);
+
+            summary.push(success);
+        }
+    }
+    summary
+}
+
```

Comment:
A more radical approach :D Warning, you might not like it :D 
Discovering projects is its own phase:
```rust
fn discover_projects(path: &Path) -> Vec<PathBuf> {
    fs::read_dir(path)
        .expect("failed to walk directory")
        .filter_map(|entry| entry.ok())
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .flat_map(|path| {
            if dir_contains_forc_manifest(&path) {
                vec![path]
            } else {
                discover_projects(&path)
            }
        })
        .collect()
}
```

Building projects returns an iterator of `TestResult`:
```rust
fn build_recursively(path: &Path) -> impl Iterator<Item = TestResult> {
    discover_projects(path).into_iter().map(|path| {
        let output = std::process::Command::new("forc")
            .args(["build", "--generate-logged-types", "--path"])
            .arg(&path)
            .output()
            .expect("failed to run `forc build` for example project");

        let compilation_result = CompilationResult {
            path,
            stdout: String::from_utf8(output.stdout).expect("Forc output is not valid utf8"),
            stderr: String::from_utf8(output.stderr).expect("Forc output is not valid utf8"),
        };

        if output.status.success() {
            TestResult::Success(compilation_result)
        } else {
            TestResult::Failure(compilation_result)
        }
    })
}
```
with test results defined as:
```rust

struct CompilationResult {
    path: PathBuf,
    stdout: String,
    stderr: String,
}

enum TestResult {
    Success(CompilationResult),
    Failure(CompilationResult),
}
```

if you implement `Display` for the result:
```rust
impl Display for TestResult {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TestResult::Success(CompilationResult { path, .. }) => {
                write!(f, "  {}: [✓] succeeded!", path.display())
            }
            TestResult::Failure(CompilationResult {
                path,
                stdout,
                stderr,
            }) => {
                write!(f, "{stdout}")?;
                write!(f, "{stderr}")?;
                write!(f, "  {}: [x] failed!", path.display())
            }
        }
    }
}
```

then you can lazily consume the iterator and lift the decision to print to stdout as high as it can go:
```rust
    let (succeeded, failed): (Vec<_>, Vec<_>) = build_recursively(&absolute_path)
        .into_iter()
        .inspect(|result| println!("{result}"))
        .partition(|result| matches! {result, TestResult::Success(_)});
```